core/arch/arm/kernel/secure_partition.c

```diff
@@ -1000,12 +1022,187 @@ const struct ts_ops sp_ops __weak __relrodata_unpaged("sp_ops") = {
 	.dump_state = sp_dump_state,
 };
 
+static TEE_Result process_sp_pkg(uint64_t sp_pkg_pa, TEE_UUID *sp_uuid)
+{
+	enum teecore_memtypes mtype = MEM_AREA_RAM_SEC;
+	struct sp_pkg_header *sp_pkg_hdr = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	tee_mm_entry_t *mm = NULL;
+	struct fip_sp *sp = NULL;
+	size_t sp_pkg_size = 0;
+	vaddr_t sp_pkg_va = 0;
+	size_t num_pages = 0;
+
+	if (!tee_pbuf_is_sec(sp_pkg_pa, SMALL_PAGE_SIZE))
+		return TEE_ERROR_GENERIC;
+
+	/* Map only the first page of SP package to parse the header */
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, SMALL_PAGE_SIZE);
+	if (!mm)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sp_pkg_va = tee_mm_get_smem(mm);
+
+	if (core_mmu_map_contiguous_pages(sp_pkg_va, sp_pkg_pa, 1, mtype)) {
+		res = TEE_ERROR_GENERIC;
+		goto err;
+	}
+
+	sp_pkg_hdr = (struct sp_pkg_header *)sp_pkg_va;
+
+	if (sp_pkg_hdr->magic != SP_PKG_HEADER_MAGIC) {
+		EMSG("Invalid SP package magic");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (sp_pkg_hdr->version != SP_PKG_HEADER_VERSION) {
+		EMSG("Invalid SP header version");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	sp_pkg_size = sp_pkg_hdr->img_offset + sp_pkg_hdr->img_size;
```

Comment:
Can this overflow?

Reply:
Both `img_offset` and `img_size` are `uint32_t`s, so I think it's not an issue. I added a check anyway, if the `sp_pkg_header` would change in the future to use 64-bit fields, it'd be easy to forget updating this line.